src/aiida/orm/nodes/data/code/installed.py

```diff
@@ -73,21 +76,25 @@ def validate_computer(cls, value: str) -> Computer:
             except exceptions.NotExistent as exception:
                 raise ValueError(exception) from exception
 
-        @model_validator(mode='after')  # type: ignore[misc]
-        def validate_full_label_uniqueness(self) -> AbstractCode.Model:
-            """Validate that the full label does not already exist."""
-            from aiida.orm import load_code
-
-            full_label = f'{self.label}@{self.computer.label}'  # type: ignore[attr-defined]
-
-            try:
-                load_code(full_label)
-            except exceptions.NotExistent:
-                return self
-            except exceptions.MultipleObjectsError as exception:
-                raise ValueError(f'Multiple codes with the label `{full_label}` already exist.') from exception
-            else:
-                raise ValueError(f'A code with the label `{full_label}` already exists.')
+        @field_serializer('computer')
+        def serialize_computer(self, computer: Computer, _info):
+            return computer.label
+
+        # @model_validator(mode='after')  # type: ignore[misc]
```

Comment:
Seems like some code that was forgotten

Reply:
Thanks for checking out this PR @agoscinski . For some background info: I commented this out on purpose for now. Ideally, this check _should_ be there because when storing the code, this condition should be satisfied. Unfortunately, however, this now also prevents just instantiating this model that validates the check, even if we don't plan on storing it. This is problematic for the export behavior, because that would rely on instantiating a `Model` that violates this uniqueness rules, even though in this case it doesn't apply.

The old CLI `verdi code create core.code.installed` would also rely on something like this check to immediately return an error as soon as `label` was specified in interactive mode and that label already existed for the same computer. It would then reprompt the user asking for a different label. I haven't been able to reimplement this feature in this PR so was keeping the comment as a reminder.

Safe to say that this PR is not completely done yet 😅

Reply:
Is this still "in progress"?

Reply:
@sphuber I'm looking at this now. Trying to probe the issue you mention in this thread. I did the following:
1. Create a `hello_world@localhost` installed code using verdi
2. Load the code and serialize it
3. Create an `InstalledCode` instance from 2
I expected an issue here, as `from_serialized` leverages the model, so the validator should've been triggered, raising an exception, as the label + computer already exist! No exception though.

As for your comment regarding the CLI, are you saying a similar validation should be implemented there? Sounds like it, but not in this PR, right?

Reply:
@sphuber what about https://docs.pydantic.dev/latest/api/base_model/#pydantic.BaseModel.model_construct for this application? It would allow us to create the model from validated data **without** triggering the validator.

Reply:
@sphuber one thing that is still unclear to us is how in the CLI application a `Model` is created from an existing `Code`. Are we not discussing the creation of codes? How is there already a code?

Reply:
```python
class A:
    class Model(BaseModel):
        a: int = Field(description='a')
        b: int = Field(description='b')

        @model_validator(mode='after')
        def validate_sum(self):
            if self.a + self.b > 10:
                raise ValueError('a + b must be less than 10')
            return self
```
`A.Model(**{"a": 5, "b": 6})` will raise a `ValueError`, but `A.Model.model_construct(**{"a": 5, "b": 6})` will not!

Reply:
> @sphuber what about https://docs.pydantic.dev/latest/api/base_model/#pydantic.BaseModel.model_construct for this application? It would allow us to create the model from validated data without triggering the validator.

This could perhaps work. We would simply change `Entity.to_model()` to return `self.Model.model_construct(**fields)` instead. In principle that would circumvent the validators since the method `model_construct` expects the data it gets to already have been validated.

However, this would cause `InstalledCode.serialize()` to except. The `validate_computer` validator is now _also_ skipped when constructing the model through `model_construct` and so the `computer` field it gets is just the `str` label and _not_ the `Computer` instance that it _would_ have had when the `validate_computer` validator is called.
```
        @field_validator('computer')
        @classmethod
        def validate_computer(cls, value: str) -> Computer:
            """Override the validator for the ``label`` of the base class since uniqueness is defined on full label."""
            from aiida.orm import load_computer

            try:
                return load_computer(value)
            except exceptions.NotExistent as exception:
                raise ValueError(exception) from exception
```
The exception is raised because when `Model.model_dump()` is then called in `Entity.serialize`, the `InstalledCode.serialize_computer` serialized method is called which calls `self.computer.label`, but `self.computer` now returns a `str` instead of a `Computer`. The problem can be immediately fixed by changing the `InstalledCode.serialize_computer` to:
```python
        @field_serializer('computer')
        def serialize_computer(self, computer: Computer, _info):
            if isinstance(computer, Computer):
                return computer.label
            return computer
```
Now the `InstalledCode.serialize()` method successfully returns a serialized version, even with the mode validator uncommented. However, I am not sure if this quick hack in `serialize_computer` has any other unintended knock-on effects.

Reply:
> @sphuber one thing that is still unclear to us is how in the CLI application a Model is created from an existing Code. Are we not discussing the creation of codes? How is there already a code?

It doesn't create a Model from an existing code. What the CLI code does is directly construct a Model from the values passed by the user to the CLI command. And then uses `InstalledCode.from_model(model)` to instantiate the `InstalledCode` instance which it then will store. The relevant line in the code:
https://github.com/sphuber/aiida-core/blob/d4c4d94851836e58e743d8f1a2d30c665af2ae3b/src/aiida/cmdline/commands/cmd_code.py#L35

So in that particular use case, the CLI implementation for creating codes, the fact that instantiating `InstalledCode.Model` for a `label` and `computer` that already exist raises a validation error is actually exactly what we want. Because we _intend_ to then store an actual `InstalledCode` instance from that model directly after. The problem only comes in that now we can never instantiate this Model _even_ if we never intend to store it as another ORM instance. The main example is when we want to serialize an existing ORM instance through the Model as an intermediate, as we have discussed in the comments above.

Reply:
Thanks @sphuber. Two things:

1. Regarding unintended effects of the `serialize_computer` guard, I would argue that any situation prior to the future merging of this PR that required the label of the computer was a situation in which the computer was indeed an instance of `Computer` and thus would not be affected by this change. The only cases where `computer` is a `str` should be the serialization case, no?
2. Regarding the CLI case...
>The problem only comes in that now we can never instantiate this Model even if we never intend to store it as another ORM instance. The main example is when we want to serialize an existing ORM instance through the Model as an intermediate, as we have discussed in the comments above.
Are you saying there IS a problem still, or are you simply reiterating the original problem? Because it seems to me that the use of `model_construct` is precisely the solution. Or am I missing something?

Reply:
> Regarding unintended effects of the serialize_computer guard, I would argue that any situation prior to the future merging of this PR that required the label of the computer was a situation in which the computer was indeed an instance of Computer and thus would not be affected by this change. The only cases where computer is a str should be the serialization case, no?

I am not worried about existing code, since the serialization is completely new. But I was just saying that with adding this workaround, although it fixes the immediate problem with using `model_construct` (which is itself already a bit of a workaround), I cannot guarantee that this will not lead to other problems/inconsistencies in the future. But perhaps this is still the best solution for now.

> Are you saying there IS a problem still, or are you simply reiterating the original problem? Because it seems to me that the use of model_construct is precisely the solution. Or am I missing something?

No I was simply answering your question and providing more explanation. Using `model_construct` in the serialization code does seem to workaround the problem for now (in combination with the guard in the computer serializer)

Reply:
Okay. Thanks for clarifying this matter 🙏

Reply:
Hmm, I think maybe there's another potentially critical issue here.

I tried reconstructing an installed code instance. With the validation active on the model, I first need to make the above change (and related changes) to `Entity.to_model()`. This allows me to serialize the code. However, deserializing it using `orm.InstalledCode.from_serialized(**code.serialize())` also raises the validation error. But here we shouldn't assume pre-validated input, so we can't use the `model_construct` solution in `from_serialized`.

I'm not sure how to get around this matter and allow the model validator. A REST API receiving a a serialized `Code` instance cannot assume the data is valid. In the case that the data came from the same AiiDA instance with which the REST API is linked, the validation will surely object.